File: frontend/src/types/Step.ts
import type { StepConnection } from './StepConnection'

export enum StepType {
  API_CALL = 'api_call',
  LLM_CALL = 'llm_call',
}

export interface Step {
  id: string
  flow_id: string
  name: string
  type: StepType
  config: StepConfig | null
  pos_x: number
  pos_y: number
  is_start: boolean
  created_at: string
  source_connections: StepConnection[]
  target_connections: StepConnection[]
  variables?: string[]
}

export interface StepCreate {
  flow_id: string
  type: StepType
  config?: StepConfig | null
  pos_x?: number
  pos_y?: number
  is_start?: boolean
  variables?: string[]
}

export interface StepUpdate {
  type?: StepType
  config?: StepConfig | null
  pos_x?: number
  pos_y?: number
  is_start?: boolean
  variables?: string[]
}

export interface StepPosition {
  pos_x: number
  pos_y: number
}

export type StepConfig = ApiNodeConfig | LlmNodeConfig

export interface ApiNodeConfig {
  apiUrl?: string
  method?: string
  data?: any
}

export interface LlmNodeConfig {
  prompt?: string
  model?: string
  temperature?: number
}

File: backend/app/schemas/step.py
from pydantic import BaseModel
import uuid
from datetime import datetime
from typing import Optional, Dict, Any

from .step_connection import StepConnection

class StepBase(BaseModel):
    name: str = "New Step"
    type: str
    config: Optional[Dict[str, Any]] = None
    pos_x: float = 0.0
    pos_y: float = 0.0
    is_start: bool = False
    variables: Optional[list[str]] = None
    mappings: Optional[list[Dict[str, str]]] = None

class StepCreate(StepBase):
    flow_id: uuid.UUID

class StepUpdate(StepBase):
    pass

class StepInDBBase(StepBase):
    id: uuid.UUID
    flow_id: uuid.UUID
    created_at: datetime

    class Config:
        from_attributes = True

class Step(StepInDBBase):
    source_connections: list[StepConnection] = []
    target_connections: list[StepConnection] = []

class StepInDB(StepInDBBase):
    pass
